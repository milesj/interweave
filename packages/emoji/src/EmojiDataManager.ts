import {
	Emoji,
	EMOTICON_OPTIONS,
	fromCodepointToUnicode,
	fromHexcodeToCodepoint,
	generateEmoticonPermutations,
	GroupKey,
	Hexcode,
	Locale,
	MessagesDataset,
	SkinToneKey,
	SubgroupKey,
	TEXT,
} from 'emojibase';
import { CanonicalEmoji } from './types';

const instances = new Map<string, EmojiDataManager>();

export function resetInstances() {
	if (__DEV__) {
		instances.clear();
	}
}

export class EmojiDataManager {
	EMOJIS: Record<string, CanonicalEmoji> = {};

	EMOTICON_TO_HEXCODE: Record<string, Hexcode> = {};

	SHORTCODE_TO_HEXCODE: Record<string, Hexcode> = {};

	UNICODE_TO_HEXCODE: Record<string, Hexcode> = {};

	GROUPS_BY_KEY: Partial<Record<GroupKey, string>> = {};

	SKIN_TONES_BY_KEY: Partial<Record<SkinToneKey, string>> = {};

	SUBGROUPS_BY_KEY: Partial<Record<SubgroupKey, string>> = {};

	data: CanonicalEmoji[] = [];

	flatData: CanonicalEmoji[] = [];

	locale: Locale = 'en';

	version: string = '0.0.0';

	constructor(locale: Locale, version: string) {
		this.locale = locale;
		this.version = version;
	}

	/**
	 * Return or create a singleton instance per locale.
	 */
	static getInstance(locale: Locale, version: string): EmojiDataManager {
		const key = `${locale}:${version}`;

		if (!instances.has(key)) {
			instances.set(key, new EmojiDataManager(locale, version));
		}

		return instances.get(key)!;
	}

	/**
	 * Return dataset as a list.
	 */
	getData(): CanonicalEmoji[] {
		return this.data;
	}

	/**
	 * Return dataset as a flattened list.
	 */
	getFlatData(): CanonicalEmoji[] {
		return this.flatData;
	}

	/**
	 * Package the emoji object with additional data,
	 * while also extracting and partitioning relevant information.
	 */
	packageEmoji(baseEmoji: Emoji): CanonicalEmoji {
		const { emoticon, hexcode, shortcodes = [] } = baseEmoji;
		const emoji: CanonicalEmoji = {
			...baseEmoji,
			canonicalShortcodes: [],
			primaryShortcode: '',
			skins: [],
			unicode: '',
		};

		// Make our lives easier
		if (!emoji.unicode) {
			emoji.unicode = emoji.text && emoji.type === TEXT ? emoji.text : emoji.emoji;
		}

		// Canonicalize the shortcodes for easy reuse
		emoji.canonicalShortcodes = shortcodes.map((code) => `:${code}:`);
		emoji.primaryShortcode = emoji.canonicalShortcodes[0];

		// Support all shortcodes
		emoji.canonicalShortcodes.forEach((shortcode) => {
			this.SHORTCODE_TO_HEXCODE[shortcode] = hexcode;
		});

		// Support all emoticons
		if (emoticon) {
			const emoticons = Array.isArray(emoticon) ? emoticon : [emoticon];

			emoticons.forEach((emo) => {
				generateEmoticonPermutations(emo, EMOTICON_OPTIONS[emo]).forEach((e) => {
					this.EMOTICON_TO_HEXCODE[e] = hexcode;
				});
			});
		}

		// Support all presentations (even no variation selectors)
		this.UNICODE_TO_HEXCODE[fromCodepointToUnicode(fromHexcodeToCodepoint(hexcode))] = hexcode;

		if (emoji.emoji) {
			this.UNICODE_TO_HEXCODE[emoji.emoji] = hexcode;
		}

		if (emoji.text) {
			this.UNICODE_TO_HEXCODE[emoji.text] = hexcode;
		}

		// Map each emoji
		this.EMOJIS[hexcode] = emoji;

		// Apply same logic to all variations
		if (baseEmoji.skins) {
			emoji.skins = baseEmoji.skins.map((skinEmoji) => this.packageEmoji(skinEmoji));
		}

		return emoji;
	}

	/**
	 * Parse and generate emoji datasets.
	 */
	parseEmojiData(data: Emoji[]): CanonicalEmoji[] {
		data.forEach((emoji) => {
			const packagedEmoji = this.packageEmoji(emoji);

			this.data.push(packagedEmoji);
			this.flatData.push(packagedEmoji);

			// Flatten and package skins as well
			if (packagedEmoji.skins) {
				packagedEmoji.skins.forEach((skin) => {
					this.flatData.push(skin);
				});
			}
		});

		return this.data;
	}

	parseMessageData(data: MessagesDataset) {
		if (data.groups) {
			data.groups.forEach((group) => {
				this.GROUPS_BY_KEY[group.key] = group.message;
			});
		}

		if (data.subgroups) {
			data.subgroups.forEach((group) => {
				this.SUBGROUPS_BY_KEY[group.key] = group.message;
			});
		}

		if (data.skinTones) {
			data.skinTones.forEach((skinTone) => {
				this.SKIN_TONES_BY_KEY[skinTone.key] = skinTone.message;
			});
		}
	}
}
