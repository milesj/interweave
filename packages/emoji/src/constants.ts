import { EmojiConfig } from './types';

// We hardcode these values so that newer emojis
// are not displayed for devices that do not support
// them yet. Support cycle is twice a year.
export const MAX_EMOJI_VERSION = 14;
export const LATEST_DATASET_VERSION = '7.0.1';

export const DEFAULT_CONFIG: EmojiConfig = {
	largeSize: '3em',
	renderUnicode: false,
	size: '1em',
};
