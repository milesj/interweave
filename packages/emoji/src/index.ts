/**
 * @copyright   2021, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

export * from './constants';
export * from './Emoji';
export * from './EmojiDataManager';
export * from './matchers';
export * from './types';
export * from './useEmojiData';
