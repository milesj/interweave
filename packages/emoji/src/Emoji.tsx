/* eslint-disable complexity */

import React, { useMemo } from 'react';
import { EmojiDataManager } from './EmojiDataManager';
import { EmojiProps, Size } from './types';

export function Emoji({
	emoticon,
	enlarge = false,
	hexcode,
	largeSize = '3em',
	path = '{{hexcode}}',
	renderUnicode = false,
	shortcode,
	size = '1em',
	source,
	unicode,
}: EmojiProps) {
	const data = EmojiDataManager.getInstance(source.locale, source.version);
	const styles = useMemo(() => {
		const styles: Record<string, Size | string> = {
			display: 'inline-block',
			verticalAlign: 'middle',
		};

		// Handle large styles
		if (enlarge && largeSize) {
			styles.width = largeSize;
			styles.height = largeSize;

			// Only apply styles if a size is defined
		} else if (size) {
			styles.width = size;
			styles.height = size;
		}

		return styles;
	}, [enlarge, size, largeSize]);

	if (__DEV__ && !emoticon && !shortcode && !unicode && !hexcode) {
		throw new Error(
			'Emoji component requires a `unicode` character, `emoticon`, `hexcode`, or a `shortcode`.',
		);
	}

	// Retrieve applicable unicode character
	let hex = hexcode;

	if (!hex && shortcode) {
		hex = data.SHORTCODE_TO_HEXCODE[shortcode];
	}

	if (!hex && emoticon) {
		hex = data.EMOTICON_TO_HEXCODE[emoticon];
	}

	if (!hex && unicode) {
		hex = data.UNICODE_TO_HEXCODE[unicode];
	}

	// Return the invalid value instead of erroring
	if (!hex || !data.EMOJIS[hex]) {
		return <span>{unicode ?? emoticon ?? shortcode ?? hex}</span>;
	}

	const emoji = data.EMOJIS[hex];

	if (renderUnicode) {
		return <span>{emoji.unicode}</span>;
	}

	// Determine the path
	let src = path || '{{hexcode}}';

	src =
		typeof src === 'function'
			? src(emoji.hexcode, {
					enlarged: enlarge,
					largeSize,
					size: enlarge ? largeSize : size,
					smallSize: size,
			  })
			: src.replace('{{hexcode}}', emoji.hexcode);

	// http://git.emojione.com/demos/latest/sprites-png.html
	// http://git.emojione.com/demos/latest/sprites-svg.html
	// https://css-tricks.com/using-svg/
	return (
		<img
			alt={emoji.unicode}
			aria-label={emoji.label}
			data-emoticon={emoji.emoticon}
			data-hexcode={emoji.hexcode}
			data-shortcodes={emoji.canonicalShortcodes.join(', ')}
			data-unicode={emoji.unicode}
			src={src}
			style={styles}
			title={emoji.label}
		/>
	);
}
