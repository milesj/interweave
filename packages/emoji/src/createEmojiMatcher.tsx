import React from 'react';
import {
	createMatcher,
	InterweaveProps,
	MatcherFactory,
	MatcherFactoryData,
	Node,
	OnMatch,
} from 'interweave';
import { DEFAULT_CONFIG } from './constants';
import { Emoji } from './Emoji';
import { EmojiConfig, EmojiMatch } from './types';

function factory({
	config,
	params,
	props: { emojiSource },
}: MatcherFactoryData<EmojiMatch, InterweaveProps, EmojiConfig>) {
	return <Emoji {...config} {...params} source={emojiSource} />;
}

function onBeforeParse(content: string, { emojiSource }: InterweaveProps): string {
	if (__DEV__ && !emojiSource) {
		throw new Error(
			'Missing emoji source data. Have you loaded with the `useEmojiData` hook and passed the `emojiSource` prop?',
		);
	}

	return content;
}

function onAfterParse(node: Node, { emojiEnlargeThreshold = 1 }: InterweaveProps): Node {
	const content = React.Children.toArray(node);

	if (content.length === 0) {
		return content;
	}

	let valid = false;
	let count = 0;

	// Use a for-loop, as it's much cleaner than some()
	for (const item of content) {
		if (typeof item === 'string') {
			// Allow whitespace but disallow strings
			if (!item.match(/^\s+$/)) {
				valid = false;
				break;
			}
		} else if (React.isValidElement(item)) {
			// Only count towards emojis
			if (item && item.type === Emoji) {
				count += 1;
				valid = true;

				if (count > emojiEnlargeThreshold) {
					valid = false;
					break;
				}

				// Abort early for non-emoji components
			} else {
				valid = false;
				break;
			}
		} else {
			valid = false;
			break;
		}
	}

	if (!valid) {
		return content;
	}

	return content.map((item) => {
		if (!React.isValidElement(item)) {
			return item;
		}

		return React.cloneElement(item, {
			...(item.props as {}),
			enlarge: true,
		});
	});
}

export function createEmojiMatcher(
	pattern: RegExp,
	onMatch: OnMatch<EmojiMatch, InterweaveProps>,
	customFactory: MatcherFactory<EmojiMatch, InterweaveProps, EmojiConfig> = factory,
) {
	return createMatcher<EmojiMatch, InterweaveProps, EmojiConfig>(pattern, customFactory, {
		config: DEFAULT_CONFIG,
		greedy: true,
		onAfterParse,
		onBeforeParse,
		onMatch,
		tagName: 'img',
		void: true,
	});
}
