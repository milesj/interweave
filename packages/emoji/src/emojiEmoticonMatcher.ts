import EMOTICON_REGEX from 'emojibase-regex/emoticon';
import EmojiDataManager from './EmojiDataManager';
import createEmojiMatcher from './createEmojiMatcher';

const EMOTICON_BOUNDARY_REGEX = new RegExp(
  // eslint-disable-next-line no-useless-escape
  `(^|\\\b|\\\s)(${EMOTICON_REGEX.source})(?=\\\s|\\\b|$)`,
);

export default createEmojiMatcher(EMOTICON_BOUNDARY_REGEX, (result, { emojiSource }) => {
  const data = EmojiDataManager.getInstance(emojiSource.locale);
  const emoticon = result.matches[0].trim();

  if (!emoticon || !data.EMOTICON_TO_HEXCODE[emoticon]) {
    return null;
  }

  // Remove padding
  // TODO still needed?
  result.match = String(emoticon);

  return {
    emoticon,
    hexcode: data.EMOTICON_TO_HEXCODE[emoticon],
  };
});
