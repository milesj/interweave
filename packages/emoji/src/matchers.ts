import EMOJI_REGEX from 'emojibase-regex';
import EMOTICON_REGEX from 'emojibase-regex/emoticon';
import SHORTCODE_REGEX from 'emojibase-regex/shortcode';
import { createEmojiMatcher } from './createEmojiMatcher';
import { EmojiDataManager } from './EmojiDataManager';

const EMOTICON_BOUNDARY_REGEX = new RegExp(
	// eslint-disable-next-line no-useless-escape
	`(^|\\\b|\\\s)(${EMOTICON_REGEX.source})(?=\\\s|\\\b|$)`,
);

export const emojiEmoticonMatcher = createEmojiMatcher(
	EMOTICON_BOUNDARY_REGEX,
	(result, { emojiSource }) => {
		const data = EmojiDataManager.getInstance(emojiSource.locale, emojiSource.version);
		const emoticon = result.matches[0].trim();

		if (!emoticon || !data.EMOTICON_TO_HEXCODE[emoticon]) {
			return null;
		}

		return {
			emoticon,
			hexcode: data.EMOTICON_TO_HEXCODE[emoticon],
			// Remove padding
			match: String(emoticon),
		};
	},
);

export const emojiShortcodeMatcher = createEmojiMatcher(
	SHORTCODE_REGEX,
	(result, { emojiSource }) => {
		const data = EmojiDataManager.getInstance(emojiSource.locale, emojiSource.version);
		const shortcode = result.matches[0].toLowerCase();

		if (!shortcode || !data.SHORTCODE_TO_HEXCODE[shortcode]) {
			return null;
		}

		return {
			hexcode: data.SHORTCODE_TO_HEXCODE[shortcode],
			shortcode,
		};
	},
);

export const emojiUnicodeMatcher = createEmojiMatcher(EMOJI_REGEX, (result, { emojiSource }) => {
	const data = EmojiDataManager.getInstance(emojiSource.locale, emojiSource.version);
	const unicode = result.matches[0];

	if (!unicode || !data.UNICODE_TO_HEXCODE[unicode]) {
		return null;
	}

	return {
		hexcode: data.UNICODE_TO_HEXCODE[unicode],
		unicode,
	};
});
