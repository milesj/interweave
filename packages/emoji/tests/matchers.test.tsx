import React from 'react';
import EMOJI_REGEX from 'emojibase-regex';
import EMOTICON_REGEX from 'emojibase-regex/emoticon';
import SHORTCODE_REGEX from 'emojibase-regex/shortcode';
import { InterweaveProps, Parser } from 'interweave';
import {
	createExpectedToken,
	parentConfig,
	SOURCE_PROP,
	TOKEN_LOCATIONS,
	VALID_EMOJIS,
} from 'interweave/test';
import { DEFAULT_CONFIG } from '../src/constants';
import { Emoji } from '../src/Emoji';
import { EmojiDataManager } from '../src/EmojiDataManager';
import { emojiEmoticonMatcher, emojiShortcodeMatcher, emojiUnicodeMatcher } from '../src/matchers';

const INVALID_UNICODE = ['\u02A9', '\u03C6', '\u0544'];

const INVALID_SHORTCODE = [':no_ending', 'no_beginning:', ':no spaces:', ':no#($*chars:'];

const INVALID_EMOTICON = ['[:', '@=', '+['];

const MAN_EMOJI = '👨';

describe('EmojiMatcher', () => {
	let EMOTICON_TO_HEXCODE: Record<string, string> = {};
	let SHORTCODE_TO_HEXCODE: Record<string, string> = {};
	let UNICODE_TO_HEXCODE: Record<string, string> = {};
	let VALID_EMOTICON: string[] = [];
	let VALID_SHORTCODE: string[] = [];
	let VALID_UNICODE: string[] = [];

	const pattern = new RegExp(`^${EMOJI_REGEX.source}$`);
	const emoPattern = new RegExp(`^${EMOTICON_REGEX.source}$`);
	const shortPattern = new RegExp(`^${SHORTCODE_REGEX.source}$`);
	const props: InterweaveProps = {
		emojiSource: SOURCE_PROP,
	};

	beforeEach(() => {
		const data = EmojiDataManager.getInstance('en', '0.0.0');

		({ EMOTICON_TO_HEXCODE, SHORTCODE_TO_HEXCODE, UNICODE_TO_HEXCODE } = data);

		VALID_EMOTICON = Object.keys(EMOTICON_TO_HEXCODE);
		VALID_SHORTCODE = Object.keys(SHORTCODE_TO_HEXCODE);
		VALID_UNICODE = Object.keys(UNICODE_TO_HEXCODE);

		// matcher.data = data;
	});

	describe('does match valid emoji', () => {
		VALID_UNICODE.forEach((unicode) => {
			// Emoji_Tag_Sequences currently do not work
			if (unicode === '🏴󠁧󠁢󠁥󠁮󠁧󠁿' || unicode === '🏴󠁧󠁢󠁳󠁣󠁴󠁿' || unicode === '🏴󠁧󠁢󠁷󠁬󠁳󠁿') {
				return;
			}

			it(`unicode: ${unicode}`, () => {
				expect(unicode.match(pattern)![0]).toBe(unicode);
			});
		});

		VALID_SHORTCODE.forEach((shortcode) => {
			it(`shortcode: ${shortcode}`, () => {
				expect(shortcode.match(shortPattern)![0]).toBe(shortcode);
			});
		});

		VALID_EMOTICON.forEach((emoticon) => {
			it(`emoticon: ${emoticon}`, () => {
				expect(emoticon.match(emoPattern)![0]).toBe(emoticon);
			});
		});
	});

	describe('doesnt match invalid emoji', () => {
		INVALID_UNICODE.forEach((unicode) => {
			it(`unicode: ${unicode}`, () => {
				expect(unicode.match(pattern)).toBeNull();
			});
		});

		INVALID_SHORTCODE.forEach((shortcode) => {
			it(`shortcode: ${shortcode}`, () => {
				expect(shortcode.match(shortPattern)).toBeNull();
			});
		});

		INVALID_EMOTICON.forEach((emoticon) => {
			it(`emoticon: ${emoticon}`, () => {
				expect(emoticon.match(emoPattern)).toBeNull();
			});
		});
	});

	describe('matches all emojis in a string', () => {
		const parser = new Parser('', props, [
			emojiUnicodeMatcher,
			emojiShortcodeMatcher,
			emojiEmoticonMatcher,
		]);

		function createUnicode(unicode: string, key: number) {
			return emojiUnicodeMatcher.factory(
				{
					config: DEFAULT_CONFIG,
					params: { hexcode: UNICODE_TO_HEXCODE[unicode], unicode },
					props,
				},
				null,
				key,
			);
		}

		function createShort(shortcode: string, key: number) {
			return emojiShortcodeMatcher.factory(
				{
					config: DEFAULT_CONFIG,
					params: { hexcode: SHORTCODE_TO_HEXCODE[shortcode], shortcode },
					props,
				},
				null,
				key,
			);
		}

		VALID_EMOJIS.forEach(([, unicode, shortcode]) => {
			TOKEN_LOCATIONS.forEach((location, i) => {
				it(`for: ${unicode} - ${location}`, () => {
					parser.keyIndex = -1; // Reset for easier testing

					const tokenString = location.replace(/{token}/g, unicode);
					const actual = parser.applyMatchers(tokenString, parentConfig);

					if (i === 0) {
						expect(actual).toBe(createExpectedToken(unicode, createUnicode, 0));
					} else {
						expect(actual).toEqual(createExpectedToken(unicode, createUnicode, i));
					}
				});

				it(`for: ${shortcode} - ${location}`, () => {
					parser.keyIndex = -1; // Reset for easier testing

					const tokenString = location.replace(/{token}/g, shortcode);
					const actual = parser.applyMatchers(tokenString, parentConfig);

					if (i === 0) {
						expect(actual).toBe(createExpectedToken(shortcode, createShort, 0));
					} else {
						expect(actual).toEqual(createExpectedToken(shortcode, createShort, i));
					}
				});
			});
		});
	});

	describe('match()', () => {
		it('returns null for invalid unicode match', () => {
			expect(emojiUnicodeMatcher.match(INVALID_UNICODE[0], props)).toBeNull();
		});

		it('returns object for valid unicode match', () => {
			expect(emojiUnicodeMatcher.match(MAN_EMOJI, props)).toEqual(
				expect.objectContaining({
					index: 0,
					length: 2,
					match: MAN_EMOJI,
					params: {
						unicode: MAN_EMOJI,
						hexcode: UNICODE_TO_HEXCODE[MAN_EMOJI],
					},
					valid: true,
					void: true,
				}),
			);
		});

		it('returns null for invalid shortcode match', () => {
			expect(emojiShortcodeMatcher.match(':invalid', props)).toBeNull();
		});

		it('returns object for valid shortcode match', () => {
			expect(emojiShortcodeMatcher.match(':man:', props)).toEqual(
				expect.objectContaining({
					index: 0,
					length: 5,
					match: ':man:',
					params: {
						shortcode: ':man:',
						hexcode: SHORTCODE_TO_HEXCODE[':man:'],
					},
					valid: true,
					void: true,
				}),
			);
		});

		it('returns null for invalid emoticon match', () => {
			expect(emojiEmoticonMatcher.match('?)', props)).toBeNull();
		});

		it('returns object for valid emoticon match', () => {
			expect(emojiEmoticonMatcher.match(':)', props)).toEqual(
				expect.objectContaining({
					index: 0,
					length: 2,
					match: ':)',
					params: {
						emoticon: ':)',
						hexcode: EMOTICON_TO_HEXCODE[':)'],
						match: ':)',
					},
					valid: true,
					void: true,
				}),
			);
		});
	});

	describe('onBeforeParse()', () => {
		it('errors when no emojiSource', () => {
			expect(() => {
				emojiUnicodeMatcher.onBeforeParse?.(
					'',
					// @ts-expect-error Missing prop
					{},
				);
			}).toThrow(
				'Missing emoji source data. Have you loaded with the `useEmojiData` hook and passed the `emojiSource` prop?',
			);
		});

		it('doesnt error when emojiSource is passed', () => {
			expect(() => {
				emojiUnicodeMatcher.onBeforeParse?.('', props);
			}).not.toThrow();
		});
	});

	describe('onAfterParse', () => {
		it('returns when an empty array', () => {
			expect(emojiUnicodeMatcher.onAfterParse?.([], props)).toEqual([]);
		});

		it('enlarges a single <Emoji/>', () => {
			expect(
				emojiUnicodeMatcher.onAfterParse?.(
					[<Emoji key={0} shortcode=":cat:" source={SOURCE_PROP} />],
					props,
				),
			).toMatchSnapshot();
		});

		it('enlarges multiple <Emoji/>s when `enlargeThreshold` is set', () => {
			expect(
				emojiUnicodeMatcher.onAfterParse?.(
					[
						<Emoji key={0} shortcode=":cat:" source={SOURCE_PROP} />,
						<Emoji key={1} shortcode=":dog:" source={SOURCE_PROP} />,
						<Emoji key={2} shortcode=":man:" source={SOURCE_PROP} />,
					],
					{
						...props,
						emojiEnlargeThreshold: 3,
					},
				),
			).toMatchSnapshot();
		});

		it('enlarge when <Emoji/> count is below `enlargeThreshold`', () => {
			expect(
				emojiUnicodeMatcher.onAfterParse?.(
					[
						<Emoji key={0} shortcode=":cat:" source={SOURCE_PROP} />,
						<Emoji key={1} shortcode=":man:" source={SOURCE_PROP} />,
					],
					{
						...props,
						emojiEnlargeThreshold: 3,
					},
				),
			).toMatchSnapshot();

			expect(
				emojiUnicodeMatcher.onAfterParse?.(
					[<Emoji key={0} shortcode=":cat:" source={SOURCE_PROP} />],
					props,
				),
			).toMatchSnapshot();
		});

		it('doesnt count whitespace in the threshold', () => {
			expect(
				emojiUnicodeMatcher.onAfterParse?.(
					[
						<Emoji key={0} shortcode=":cat:" source={SOURCE_PROP} />,
						' ',
						<Emoji key={1} shortcode=":dog:" source={SOURCE_PROP} />,
						'\n',
						<Emoji key={2} shortcode=":man:" source={SOURCE_PROP} />,
					],
					{
						...props,
						emojiEnlargeThreshold: 3,
					},
				),
			).toMatchSnapshot();
		});

		it('doesnt enlarge when too many <Emoji/>', () => {
			expect(
				emojiUnicodeMatcher.onAfterParse?.(
					[
						<Emoji key={0} shortcode=":cat:" source={SOURCE_PROP} />,
						<Emoji key={1} shortcode=":dog:" source={SOURCE_PROP} />,
						<Emoji key={2} shortcode=":man:" source={SOURCE_PROP} />,
						<Emoji key={3} shortcode=":woman:" source={SOURCE_PROP} />,
					],
					{
						...props,
						emojiEnlargeThreshold: 3,
					},
				),
			).toMatchSnapshot();
		});

		it('doesnt enlarge when strings are found', () => {
			expect(
				emojiUnicodeMatcher.onAfterParse?.(
					[
						<Emoji key={0} shortcode=":cat:" source={SOURCE_PROP} />,
						'Foo',
						<Emoji key={2} shortcode=":man:" source={SOURCE_PROP} />,
					],
					{
						...props,
						emojiEnlargeThreshold: 3,
					},
				),
			).toMatchSnapshot();
		});

		it('doesnt enlarge when non-<Emoji/> are found', () => {
			expect(
				emojiUnicodeMatcher.onAfterParse?.(
					[
						<Emoji key={0} shortcode=":cat:" source={SOURCE_PROP} />,
						<div key="foo">Foo</div>,
						<Emoji key={2} shortcode=":man:" source={SOURCE_PROP} />,
					],
					{
						...props,
						emojiEnlargeThreshold: 3,
					},
				),
			).toMatchSnapshot();
		});

		it('ignores non-<Emoji/>', () => {
			expect(
				emojiUnicodeMatcher.onAfterParse?.([<div key="foo">Foo</div>], {
					...props,
					emojiEnlargeThreshold: 3,
				}),
			).toMatchSnapshot();
		});

		it('ignores content longer than 1', () => {
			expect(
				emojiUnicodeMatcher.onAfterParse?.([<div key="foo">Foo</div>, 'Bar'], {
					...props,
					emojiEnlargeThreshold: 3,
				}),
			).toMatchSnapshot();
		});
	});
});
