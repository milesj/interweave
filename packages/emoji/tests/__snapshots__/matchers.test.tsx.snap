// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmojiMatcher onAfterParse doesnt count whitespace in the threshold 1`] = `
Array [
  <Emoji
    enlarge={true}
    shortcode=":cat:"
    source={
      Object {
        "compact": false,
        "locale": "en",
        "version": "0.0.0",
      }
    }
  />,
  " ",
  <Emoji
    enlarge={true}
    shortcode=":dog:"
    source={
      Object {
        "compact": false,
        "locale": "en",
        "version": "0.0.0",
      }
    }
  />,
  "
",
  <Emoji
    enlarge={true}
    shortcode=":man:"
    source={
      Object {
        "compact": false,
        "locale": "en",
        "version": "0.0.0",
      }
    }
  />,
]
`;

exports[`EmojiMatcher onAfterParse doesnt enlarge when non-<Emoji/> are found 1`] = `
Array [
  <Emoji
    shortcode=":cat:"
    source={
      Object {
        "compact": false,
        "locale": "en",
        "version": "0.0.0",
      }
    }
  />,
  <div>
    Foo
  </div>,
  <Emoji
    shortcode=":man:"
    source={
      Object {
        "compact": false,
        "locale": "en",
        "version": "0.0.0",
      }
    }
  />,
]
`;

exports[`EmojiMatcher onAfterParse doesnt enlarge when strings are found 1`] = `
Array [
  <Emoji
    shortcode=":cat:"
    source={
      Object {
        "compact": false,
        "locale": "en",
        "version": "0.0.0",
      }
    }
  />,
  "Foo",
  <Emoji
    shortcode=":man:"
    source={
      Object {
        "compact": false,
        "locale": "en",
        "version": "0.0.0",
      }
    }
  />,
]
`;

exports[`EmojiMatcher onAfterParse doesnt enlarge when too many <Emoji/> 1`] = `
Array [
  <Emoji
    shortcode=":cat:"
    source={
      Object {
        "compact": false,
        "locale": "en",
        "version": "0.0.0",
      }
    }
  />,
  <Emoji
    shortcode=":dog:"
    source={
      Object {
        "compact": false,
        "locale": "en",
        "version": "0.0.0",
      }
    }
  />,
  <Emoji
    shortcode=":man:"
    source={
      Object {
        "compact": false,
        "locale": "en",
        "version": "0.0.0",
      }
    }
  />,
  <Emoji
    shortcode=":woman:"
    source={
      Object {
        "compact": false,
        "locale": "en",
        "version": "0.0.0",
      }
    }
  />,
]
`;

exports[`EmojiMatcher onAfterParse enlarge when <Emoji/> count is below \`enlargeThreshold\` 1`] = `
Array [
  <Emoji
    enlarge={true}
    shortcode=":cat:"
    source={
      Object {
        "compact": false,
        "locale": "en",
        "version": "0.0.0",
      }
    }
  />,
  <Emoji
    enlarge={true}
    shortcode=":man:"
    source={
      Object {
        "compact": false,
        "locale": "en",
        "version": "0.0.0",
      }
    }
  />,
]
`;

exports[`EmojiMatcher onAfterParse enlarge when <Emoji/> count is below \`enlargeThreshold\` 2`] = `
Array [
  <Emoji
    enlarge={true}
    shortcode=":cat:"
    source={
      Object {
        "compact": false,
        "locale": "en",
        "version": "0.0.0",
      }
    }
  />,
]
`;

exports[`EmojiMatcher onAfterParse enlarges a single <Emoji/> 1`] = `
Array [
  <Emoji
    enlarge={true}
    shortcode=":cat:"
    source={
      Object {
        "compact": false,
        "locale": "en",
        "version": "0.0.0",
      }
    }
  />,
]
`;

exports[`EmojiMatcher onAfterParse enlarges multiple <Emoji/>s when \`enlargeThreshold\` is set 1`] = `
Array [
  <Emoji
    enlarge={true}
    shortcode=":cat:"
    source={
      Object {
        "compact": false,
        "locale": "en",
        "version": "0.0.0",
      }
    }
  />,
  <Emoji
    enlarge={true}
    shortcode=":dog:"
    source={
      Object {
        "compact": false,
        "locale": "en",
        "version": "0.0.0",
      }
    }
  />,
  <Emoji
    enlarge={true}
    shortcode=":man:"
    source={
      Object {
        "compact": false,
        "locale": "en",
        "version": "0.0.0",
      }
    }
  />,
]
`;

exports[`EmojiMatcher onAfterParse ignores content longer than 1 1`] = `
Array [
  <div>
    Foo
  </div>,
  "Bar",
]
`;

exports[`EmojiMatcher onAfterParse ignores non-<Emoji/> 1`] = `
Array [
  <div>
    Foo
  </div>,
]
`;
